### PR TITLE
Integrate the existing JS controllers into Deck component

### PR DESCRIPTION
**Pull request by ibgreen**
For #1429
#### Background
- Similar logic to the scripting PR, but now shared inside `Deck` component
#### Change List
- Add handling of existing controllers to `Deck` component

Depends on one or two of the PRs opened today, but isolated to simplify review.

### CODE REVIEW

File: src/core/lib/deck.js [hunk 1]
```diff
@@ -172,6 +183,20 @@ export default class Deck {
     return canvas;
   }
 
+  // Note: props.controller must be a class, not an already created instance
+  _createController(props) {
+    const Controller = props.controller;
```
Comment (Pessimistress) [inline src/core/lib/deck.js]: `controller` is not listed in propTypes

Comment (ibgreen) [reply src/core/lib/deck.js]: > controller is not listed in propTypes

Fixed.

File: src/core/lib/deck.js [hunk 2]
```diff
@@ -112,15 +136,9 @@ export default class Deck {
...
 
-  finalize() {
-    this.animationLoop.stop();
-    this.animationLoop = null;
-
-    if (this.layerManager) {
-      this.layerManager.finalize();
-      this.layerManager = null;
+    if (this.controller) {
+      this.controller.setProps(props);
```
Comment (Pessimistress) [inline src/core/lib/deck.js]: Need to flatten `viewState`?

Comment (ibgreen) [reply src/core/lib/deck.js]: > Need to flatten viewState?

`viewport-controls.js`, line 144 should handle it?

**Comment by ibgreen** [comment]
Yay...

**Pull request merged**